neuralop/layers/attention_kernel_integral.py

```diff
@@ -0,0 +1,212 @@
+            if param.ndim > 1:
+                for h in range(self.n_heads):
+                    inif_fn(param[h * self.head_n_channels:(h + 1) * self.head_n_channels, :], gain=self.init_gain)
+                    if self.head_n_channels == self.channels:
+                        diagonal_bias = self.diagonal_weight * torch.diag(torch.ones(param.size(-1), dtype=torch.float32))
+                        param.data[h * self.head_n_channels:(h + 1) * self.head_n_channels, :] += diagonal_bias
+
+    def initialize_qkv_weights(self):
+        """
+        Initialize the weights for q, k, v projection matrix with a small gain and add a diagonal bias,
+        this technique has been found useful for scale-sensitive problem that has not been normalized
+        see Table 8 in https://arxiv.org/pdf/2105.14995.pdf
+        """
+        init_fn = xavier_uniform_
+
+        if self.project_query:
+            self.init_weight(self.to_q, init_fn)
+        self.init_weight(self.to_k, init_fn)
+        self.init_weight(self.to_v, init_fn)
+
+    def normalize_wrt_domain(self, u, norm_fn):
+        """
+        Normalize the input function with respect to the domain,
+         reshape the tensor to [batch_size*n_heads, num_grid_points, head_n_channels]
+        The second dimension is equal to the number of grid points that discretize the domain
+        """
+        # u: the input or transformed function
+        batch_size = u.shape[0]
+        u = u.view(batch_size*self.n_heads, -1, self.head_n_channels)
+        u = norm_fn(u)    # layer norm with channel dimension or instance norm with spatial dimension
+        return u.view(batch_size, self.n_heads, -1, self.head_n_channels)
+
+    def forward(self,
+                u_x,
+                pos_x,
+                positional_embedding_module=None,    # positional encoding module for encoding q/k
+                u_y=None,
+                pos_y=None,
+                weights=None,
+                associative=True,   # can be much faster if num_grid_points is larger than the channel number c
+                return_kernel=False):
+        """
+        Computes kernel integral transform with attention
+
+        Parameters
+        ----------
+        u_x: input (query) function of shape [batch_size, num_grid_points, channels]
+        pos_x: coordinate of input function's grid points [batch_size, num_grid_points, pos_dim]
+        positional_embedding_module: positional embedding module for encoding query/key (q/k), a torch.nn.Module
+        u_y: the second source of function (key and value), if not provided, u_y = u_x
+        pos_y: coordinate of the second source of function's grid points, if not provided, assume pos_y = pos_x
+        weights : tensor of shape [batch_size, num_grid_points], if not provided assume to be 1/num_grid_points
+                  Weights for each point y proportional to the
+                  volume around f(y)=u_y W_v being integrated.
```

Comment:
It is unclear what `W_v` here is.

Reply:
Elaborated a bit more in the new docstring, it's a linear layer projecting the input to value in attention